certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,215 @@
+function safeIsModuleEnabled(address callee, address module) returns bool {
+    assert callee == safeContract;
+    return safeContract.isModuleEnabled(module);
+}
+
+// A summary function that asserts that all `ISafe.isOwner` calls are done
+// to the `safeContract`, returning the same result as `safeContract.isOwner(...)`.
+function sumarizeSafeIsOwner(address callee, address owner) returns bool {
+    assert callee == safeContract;
+    return safeContract.isOwner(owner);
+}
+
+// A setup function that requires Safe contract to enable the Social Recovery Module.
 function requireSocialRecoveryModuleEnabled() {
     require(safeContract.isModuleEnabled(currentContract));
 }
 
-// This is a dummy rule to verify the Safe Contract Setup with the Social Recovery
-// Module is working as intended.
-rule recoveryModuleCanBeDisabled {
-    env e;
-    address prevModule;
+function requireGuardiansLinkedListIntegrity() {
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+}
+
+// This integrity rule verifies that if the addGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the new guardian is added to the guardian list,
+// - and no other account (guardian or not) is affected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherAccount) {
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+
+    requireGuardiansLinkedListIntegrity();
+
+    require guardian != otherAccount;
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+// Some of the conditions are:
+// - No value should be sent with the transaction.
+// - The threshold should be greater than 0. We can't have guardian(s) with a threshold of 0.
+// - The guardian count should be less than the maximum value to prevent overflow.
+// - The guardian should not be values such as zero, sentinel, or the Safe contract itself.
+// - The guardian should not be an owner of the Safe contract at the time of addition.
+// - The guardian should not be already added as a guardian.
+// - The threshold should be less than or equal to the total number of guardians.
+// - Safe contract should be the sender of the transaction.
```

Comment:
nit: I would document these inline - this makes it harder for there to be documentation drift.

Reply:
Done 👍🏾